File: harness/tests/fake-agent.py
#!/usr/bin/env python3
"""A scripted ACP agent for bot-harness's tests (tests/fake_agent.rs).

FAKE_MODE picks what it does in session/prompt:
- fs: sends fs/* and terminal/* requests (session-scoped and not) that the
  harness doesn't advertise, and expects each answered;
- flood: sends FAKE_UPDATES agent_message_chunk notifications;
- grace: makes three tool calls, then asks permission after the harness
  must have cancelled the session (run it with --max-tool-calls 2);
- slow-init: never answers initialize.
It ends the prompt turn with end_turn (or cancelled, after session/cancel).
"""
import json
import os
import select
import sys
import time

MODE = os.environ.get("FAKE_MODE", "fs")
SESSION = "s1"


def send(m):
    sys.stdout.write(json.dumps(m) + "\n")
    sys.stdout.flush()


def recv(timeout=None):
    r, _, _ = select.select([sys.stdin], [], [], timeout)
    if not r:
        return None
    line = sys.stdin.readline()
    if not line:
        sys.exit(0)
    return json.loads(line)


def wait_for(id_, timeout):
    """The response to request ID_; notifications in between are noted."""
    end = time.time() + timeout
    cancelled = False
    while time.time() < end:
        m = recv(end - time.time())
        if m is None:
            break
        if m.get("method") == "session/cancel":
            cancelled = True
        elif m.get("id") == id_ and "method" not in m:
            return m, cancelled
    sys.exit(f"fake-agent: no response to request {id_}")


def update(u):
    send({"jsonrpc": "2.0", "method": "session/update", "params": {"sessionId": SESSION, "update": u}})


while True:
    m = recv()
    method = m.get("method")
    if method == "initialize":
        if MODE == "slow-init":
            continue
        send({"jsonrpc": "2.0", "id": m["id"], "result": {"protocolVersion": 1, "agentInfo": {"name": "fake", "version": "0"}}})
    elif method == "session/new":
        send({"jsonrpc": "2.0", "id": m["id"], "result": {"sessionId": SESSION}})
    elif method == "session/prompt":
        stop = "end_turn"
        if MODE == "fs":
            for i, (req, params) in enumerate([
                ("fs/read_text_file", {"sessionId": SESSION, "path": "/etc/passwd"}),
                ("terminal/create", {"sessionId": SESSION, "command": "id"}),
                ("fs/read_text_file", {"path": "/etc/passwd"}),
            ]):
                send({"jsonrpc": "2.0", "id": 100 + i, "method": req, "params": params})
                wait_for(100 + i, 5)
        elif MODE == "flood":
            chunk = "x" * 1000
            for _ in range(int(os.environ["FAKE_UPDATES"])):
                update({"sessionUpdate": "agent_message_chunk", "content": {"type": "text", "text": chunk}})
        elif MODE == "grace":
            for i in range(3):
                update({"sessionUpdate": "tool_call", "toolCallId": f"t{i}", "title": "Bash", "kind": "execute",
                        "status": "pending", "rawInput": {"command": "true"}})
            time.sleep(1)
            send({"jsonrpc": "2.0", "id": 300, "method": "session/request_permission", "params": {
                "sessionId": SESSION,
                "toolCall": {"toolCallId": "t2", "kind": "execute", "rawInput": {"command": "true"}},
                "options": [{"optionId": "ok", "name": "Allow", "kind": "allow_once"},
                            {"optionId": "no", "name": "Reject", "kind": "reject_once"}]}})
            _, cancelled = wait_for(300, 10)
            stop = "cancelled" if cancelled else stop
        send({"jsonrpc": "2.0", "id": m["id"], "result": {"stopReason": stop}})

File: harness/tests/fake_agent.rs
//! bot-harness against a scripted fake agent (tests/fake-agent.py), for
//! protocol behavior the real agents don't readily show.

use serde_json::Value;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{Duration, Instant};

const AGENTS: &str = "[fake]\ncommand = [\"python3\", \"tests/fake-agent.py\"]\n";

struct Run {
    status: i32,
    elapsed: Duration,
    result: Value,
    /// acp.jsonl
    records: Vec<Value>,
}

fn run(mode: &str, env: &[(&str, &str)], args: &[&str]) -> Run {
    let dir = tempfile::tempdir().unwrap();
    let path = |name: &str| -> PathBuf { dir.path().join(name) };
    std::fs::write(path("agents.toml"), AGENTS).unwrap();
    std::fs::write(path("prompt.md"), "Do the fake task.\n").unwrap();
    let root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let start = Instant::now();
    let out = Command::new(env!("CARGO_BIN_EXE_bot-harness"))
        .current_dir(root)
        .args(["run", "--agent", "fake", "--cwd", "."])
        .arg("--agents")
        .arg(path("agents.toml"))
        .arg("--prompt")
        .arg(path("prompt.md"))
        .arg("--out")
        .arg(path("out"))
        .args(args)
        .env("FAKE_MODE", mode)
        .envs(env.iter().copied())
        .output()
        .unwrap();
    let elapsed = start.elapsed();
    let read = |name: &str| std::fs::read_to_string(path("out").join(name)).unwrap();
    let result: Value = serde_json::from_str(&read("harness.json")).unwrap();
    let records = read("acp.jsonl")
        .lines()
        .map(|l| serde_json::from_str(l).unwrap())
        .collect();
    Run {
        status: out.status.code().unwrap(),
        elapsed,
        result,
        records,
    }
}

/// The harness's response to the agent's request ID.
fn response_to(records: &[Value], id: u64) -> &Value {
    records
        .iter()
        .find(|r| r["dir"] == "send" && r["msg"]["id"] == id && r["msg"]["method"].is_null())
        .unwrap_or_else(|| panic!("no response to request {id}"))
}

#[test]
fn unadvertised_requests_are_refused() {
    // Each would hang for the fake agent's 5s if left unanswered.
    let r = run("fs", &[], &["--timeout", "60s"]);
    assert_eq!(
        (r.status, r.result["result"].as_str()),
        (0, Some("success")),
        "{:#}",
        r.result
    );
    for id in [100, 101, 102] {
        assert_eq!(
            response_to(&r.records, id)["msg"]["error"]["code"],
            -32601,
            "request {id}"
        );
    }
    assert!(r.elapsed < Duration::from_secs(5), "took {:?}", r.elapsed);
}

#[test]
fn update_flood() {
    // A smoke test: unhandled, these piled up in the SDK's retry queue,
    // which shows in memory (about 1 GB for 200k), not in the result.
    let r = run(
        "flood",
        &[("FAKE_UPDATES", "20000")],
        &["--timeout", "120s"],
    );
    assert_eq!(
        (r.status, r.result["result"].as_str()),
        (0, Some("success")),
        "{:#}",
        r.result
    );
    assert!(r.records.len() > 20000);
}

#[test]
fn permission_after_cancel() {
    let r = run("grace", &[], &["--timeout", "60s", "--max-tool-calls", "2"]);
    assert_eq!(
        (r.status, r.result["result"].as_str()),
        (3, Some("budget")),
        "{:#}",
        r.result
    );
    let answer = &response_to(&r.records, 300)["msg"]["result"];
    assert_eq!(answer["outcome"]["outcome"], "cancelled", "{answer}");
    assert_eq!(answer["_meta"]["botHarness"]["decision"], "cancelled");
    // The cancel went out before the request came in.
    let cancel = r
        .records
        .iter()
        .position(|r| r["msg"]["method"] == "session/cancel")
        .unwrap();
    let request = r
        .records
        .iter()
        .position(|r| r["dir"] == "recv" && r["msg"]["id"] == 300)
        .unwrap();
    assert!(cancel < request);
}

#[test]
fn timeout_before_the_session() {
    let r = run("slow-init", &[], &["--timeout", "2s"]);
    assert_eq!(
        (r.status, r.result["result"].as_str()),
        (124, Some("timeout")),
        "{:#}",
        r.result
    );
    assert!(r.elapsed < Duration::from_secs(10), "took {:?}", r.elapsed);
}
